servicetalk-transport-api/src/main/java/io/servicetalk/transport/api/ConnectionObserver.java

```diff
@@ -28,6 +28,21 @@
  */
 public interface ConnectionObserver {
 
+    /**
+     * Callback when a connection starts initialization and {@link ConnectionInfo} becomes available.
+     * <p>
+     * This callback is invoked before any other callback on this observer.
+     *
+     * @param info {@link ConnectionInfo} for the connection that starts initialization. Note that the
+     * {@link ConnectionInfo#sslSession()} will always return {@code null} since it is called before the
+     * {@link ConnectionObserver#onSecurityHandshake() security handshake} is performed (and as a result no SSL session
+     * has been established). Also, {@link ConnectionInfo#protocol()} will return L4 (transport) protocol.
+     * Finalized {@link ConnectionInfo} will be available via {@link #connectionEstablished(ConnectionInfo)} or
+     * {@link #multiplexedConnectionEstablished(ConnectionInfo)} callbacks.
+     */
+    default void onConnectionInitialization(ConnectionInfo info) {  // FIXME: 0.43 - consider removing default impl
```

Comment:
Context for others from our offline discussion:

> I thought about that as well, but because `EarlyConnectionAcceptor` already uses `ConnectionInfo` and it's a `FunctionalInterface`, didn't want to create inconsistency. Propagation of `ExecutionContext` makes it harder indeed. I was less worried, bcz officially it's "internal" API, changes for initializers should not be visible for users in practice.